### PR TITLE
Fixing results dir

### LINKED ISSUE #14: Fix results directory behavior
@wjcunningham7 commented on [Fri Jan 21 2022](https://github.com/AgnostiqHQ/covalent-staging/issues/354)

Currently `write_streams_to_file` will attach the stdout and stderr files to the results directory if those files are specified using relative paths. However, this always queries the results directory in the config file when instead it should be looking at the electron metadata.

Edit:

The stdout/stderr log files that are written to is determined by priority. 

1. Top priority goes to the stdout/stderr log files in the covalent.conf file IF it is an absolute path. At the moment these files are overwritten but ideally they should be appended to.
2. In the case the the stdout/etderr log files in the covalent.conf file is a relative path, we first check if results_dir has been specified in the lattice metadata. If it has been specified, that has the second highest priority. The logfiles should be written to results_dir/dispatch_id/ folder.
3. Lastly, in the case where the log file paths are relative and results_dir has not been specified for electrons, we should use the results_dir in the covalent.conf file and store the log files in the results_dir/dispatch_id folder.

In this issue, we ensure that priority 2 is indeed being checked and implemented.

### PR DESCRIPTION
**Pull request by wjcunningham7**
fixes #14 

See https://github.com/AgnostiqHQ/covalent-staging/pull/371 for earlier discussion.

### CODE REVIEW

**Comment by FyzHsn** [comment]
I propose modifying the testing strategy for this bug fix. Rather than creating a workflow on the client side and then dispatching it with different stdout/stderr and specifying the `results_dir` via the lattice decorator configurations, one could instead take a more unit testing approach. The files that were changed to fix this bug are as follows:

- [ ] `_plan_workflow ` in `covalent_dispatcher/_core/__init__.py` - `result_object.results_dir` is now being passed in the `backend_args` variable.
- [ ] `get_executor` in `covalent/executor/__init__.py` - The config file is reloaded before returning the `executor_plugin_map` attribute.
- [ ] `write_streams_to_file` in `covalent/executor/base.py` - This now accepts `results_dir` as a parameter.
- [ ] `LocalExecutor.execute` in `covalent/executor/local.py` - This now passes the `results_dir` in `execution_args` passed as a parameter in the `execute` function.

Writing a unit test for each of these functions to ensure that these changes are adequately reflected will ensure code stability while allowing us to more easily mock the values in `covalent.conf`.

**Comment by wjcunningham7** [comment]
@FyzHsn I propose creating new issues for testing (blocked by https://github.com/AgnostiqHQ/covalent-staging/issues/381) so that we don't greatly increase the scope of this issue.

**Comment by FyzHsn** [comment]
> @FyzHsn I propose creating new issues for testing (blocked by [AgnostiqHQ/covalent-staging#381](https://github.com/AgnostiqHQ/covalent-staging/issues/381)) so that we don't greatly increase the scope of this issue.

@wjcunningham7 I was going to add some ad-hoc tests for this one and then use [AgnostiqHQ/covalent-staging#381](https://github.com/AgnostiqHQ/covalent-staging/issues/381) further down the line for a more thought out refactor and testing.

**Comment by wjcunningham7** [comment]
> > @FyzHsn I propose creating new issues for testing (blocked by [AgnostiqHQ/covalent-staging#381](https://github.com/AgnostiqHQ/covalent-staging/issues/381)) so that we don't greatly increase the scope of this issue.
> 
> @wjcunningham7 I was going to add some ad-hoc tests for this one (to avoid increasing the scope of this issue) and then use [AgnostiqHQ/covalent-staging#381](https://github.com/AgnostiqHQ/covalent-staging/issues/381) further down the line for a more thought out refactor and testing.

that works for me

**Comment by FyzHsn** [comment]
@wjcunningham7 I'm done with the tests for this branch. I've added @kessler-frost as a reviewer. 

I found that it's best to keep things simple and write tests to cover the lines of diff that were added. My initial suggested protocol of writing "a single" test to expose a bug simply would have been too convoluted. In this particular case, it is better to write multiple tests to catch the sub-issues that were causing the bug.

**Comment by FyzHsn** [comment]
Reminder: Open up an issue asking to fill in the type hints for the `write_stream_to_file` method with the `good first issue` tag.

File: tests/functional_tests/basic_dispatcher_test.py [hunk 1]
```diff
@@ -97,3 +103,5 @@ def outer_lattice(y):
     output = ct.get_result(dispatch_id, wait=True).result
 
     assert output == 25
+
+    shutil.rmtree(lattice_square.get_metadata(name="results_dir"), ignore_errors=True)
```
Comment (kessler-frost) [inline tests/functional_tests/basic_dispatcher_test.py:107]: `_delete_result` should take care of this as well.

**Review by kessler-frost — COMMENTED** [review]
Some comments for minor things. Functionally there doesn't seem to be a problem. Good job @FyzHsn!

File: tests/covalent_dispatcher_tests/dispatcher_stack_test.py [hunk 2]
```diff
@@ -66,3 +68,4 @@ def test_dispatcher_flow(mock_result, expected_res, expected_node_outputs):
     assert result.get_all_node_outputs() == expected_node_outputs
 
     rm._delete_result(dispatch_id=dispatch_id, results_dir=TEST_RESULTS_DIR)
+    shutil.rmtree(TEST_RESULTS_DIR, ignore_errors=True)
```
[the remark this answers is not in the document]
Comment (kessler-frost) [reply tests/covalent_dispatcher_tests/dispatcher_stack_test.py:71]: Still using `shutil` here sir 😅 .

File: tests/covalent_dispatcher_tests/init_test.py [hunk 3]
```diff
@@ -22,6 +22,8 @@
 Tests for self-contained entry point for the dispatcher
 """
 
+import shutil
```
Comment (kessler-frost) [inline tests/covalent_dispatcher_tests/init_test.py:25]: Unused import.

File: tests/functional_tests/basic_dispatcher_test.py [hunk 4]
```diff
@@ -19,7 +19,11 @@
 # Relief from the License may be granted by purchasing a commercial license.
 
 
+import shutil
```
Comment (kessler-frost) [inline tests/functional_tests/basic_dispatcher_test.py:22]: Unused import.

**Review by kessler-frost — COMMENTED** [review]
Missed unused `shutil` in some places.

File: CHANGELOG.md [hunk 5]
```diff
@@ -5,6 +5,20 @@ All notable changes to this project will be documented in this file.
...
 
+## [0.22.13] - 2022-01-26
+
+### Fixed
+
+- When the local executor's `log_stdout` and `log_stderr` config variables are relative paths, they should go inside the results directory. Previously that was queried from the config, but now it's queried from the lattice metadata.
+
+### Added
+
+- Tests for the corresponding functions in (`covalent_dispatcher/_core?__init__.py`, `covalent/executor/base.py`, `covalent/executor/executor_plugins/local.py` and `covalent/executor/__init__.py`) affected by the bug fix.
```
Comment (kessler-frost) [inline CHANGELOG.md:16]: Replace `?` with `/`.

File: CHANGELOG.md [hunk 6]
```diff
@@ -5,6 +5,20 @@ All notable changes to this project will be documented in this file.
...
+
+### Changed
+
+- Refactored `_delete_result` in result manager to allow forcefully deleting results directory.
```
Comment (kessler-frost) [inline CHANGELOG.md:20]: Hmm, not "to allow deleting the parent directory if desired."? We haven't changed the mechanism of how it is deleting the files, we have only added an option to delete another directory right?

**Review by kessler-frost — APPROVED** [review]

**Pull request merged**